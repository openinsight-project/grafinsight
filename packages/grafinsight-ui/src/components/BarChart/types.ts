import { VizOrientation } from '@grafinsight/data';
import { AxisConfig, GraphGradientMode, HideableFieldConfig } from '../uPlot/config';
import { VizLegendOptions } from '../VizLegend/types';

/**
 * @alpha
 */
export enum BarStackingMode {
  None = 'none',
  Standard = 'standard',
  Percent = 'percent',
}

/**
 * @alpha
 */
export enum BarValueVisibility {
  Auto = 'auto',
  Never = 'never',
  Always = 'always',
}

/**
 * @alpha
 */
export interface BarChartOptions {
  orientation: VizOrientation;
  legend: VizLegendOptions;
  stacking: BarStackingMode;
  showValue: BarValueVisibility;
  barWidth: number;
  groupWidth: number;
}

/**
 * @alpha
 */
export interface BarChartFieldConfig extends AxisConfig, HideableFieldConfig {
  lineWidth?: number; // 0
  fillOpacity?: number; // 100
  gradientMode?: GraphGradientMode;
}

/**
 * @alpha
 */
export const defaultBarChartFieldConfig: BarChartFieldConfig = {
  lineWidth: 1,
  fillOpacity: 80,
  gradientMode: GraphGradientMode.None,
  axisSoftMin: 0,
};
