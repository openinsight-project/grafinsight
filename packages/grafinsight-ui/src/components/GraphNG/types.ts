import { DataFrameFieldIndex, FieldMatcher } from '@grafinsight/data';

/**
 * Mode to describe if a legend is isolated/selected or being appended to an existing
 * series selection.
 * @alpha
 */
export enum GraphNGLegendEventMode {
  ToggleSelection = 'select',
  AppendToSelection = 'append',
}

/**
 * Event being triggered when the user interact with the Graph legend.
 * @alpha
 */
export interface GraphNGLegendEvent {
  fieldIndex: DataFrameFieldIndex;
  mode: GraphNGLegendEventMode;
}

export interface XYFieldMatchers {
  x: FieldMatcher; // first match
  y: FieldMatcher;
}
