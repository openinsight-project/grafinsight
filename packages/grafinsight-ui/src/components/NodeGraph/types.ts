import { SimulationNodeDatum, SimulationLinkDatum } from 'd3-force';

export type NodeDatum = SimulationNodeDatum & {
  id: string;
  title: string;
  subTitle: string;
  dataFrameRowIndex: number;
  incoming: number;
  mainStat: string;
  secondaryStat: string;
  arcSections: Array<{
    value: number;
    color: string;
  }>;
};
export type EdgeDatum = SimulationLinkDatum<NodeDatum> & {
  id: string;
  mainStat: string;
  secondaryStat: string;
  dataFrameRowIndex: number;
};
