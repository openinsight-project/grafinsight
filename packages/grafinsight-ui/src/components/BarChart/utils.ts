import { UPlotConfigBuilder } from '../uPlot/config/UPlotConfigBuilder';
import {
  DataFrame,
  FieldType,
  formattedValueToString,
  getFieldColorModeForField,
  getFieldDisplayName,
  getFieldSeriesColor,
  GrafInsightTheme,
  MutableDataFrame,
  VizOrientation,
} from '@grafinsight/data';
import { BarChartFieldConfig, BarChartOptions, BarValueVisibility, defaultBarChartFieldConfig } from './types';
import { AxisPlacement, ScaleDirection, ScaleDistribution, ScaleOrientation } from '../uPlot/config';
import { BarsOptions, getConfig } from './bars';
import { FIXED_UNIT } from '../GraphNG/GraphNG';

/** @alpha */
export function preparePlotConfigBuilder(
  data: DataFrame,
  theme: GrafInsightTheme,
  { orientation, showValue, groupWidth, barWidth }: BarChartOptions
) {
  const builder = new UPlotConfigBuilder();

  // bar orientation -> x scale orientation & direction
  let xOri = ScaleOrientation.Vertical;
  let xDir = ScaleDirection.Down;
  let yOri = ScaleOrientation.Horizontal;
  let yDir = ScaleDirection.Right;

  if (orientation === VizOrientation.Vertical) {
    xOri = ScaleOrientation.Horizontal;
    xDir = ScaleDirection.Right;
    yOri = ScaleOrientation.Vertical;
    yDir = ScaleDirection.Up;
  }

  const formatValue =
    showValue !== BarValueVisibility.Never
      ? (seriesIdx: number, value: any) => formattedValueToString(data.fields[seriesIdx].display!(value))
      : undefined;

  // Use bar width when only one field
  if (data.fields.length === 2) {
    groupWidth = barWidth;
    barWidth = 1;
  }

  const opts: BarsOptions = {
    xOri,
    xDir,
    groupWidth,
    barWidth,
    formatValue,
    onHover: (seriesIdx: number, valueIdx: number) => {
      console.log('hover', { seriesIdx, valueIdx });
    },
    onLeave: (seriesIdx: number, valueIdx: number) => {
      console.log('leave', { seriesIdx, valueIdx });
    },
  };

  const config = getConfig(opts);

  builder.addHook('init', config.init);
  builder.addHook('drawClear', config.drawClear);
  builder.addHook('setCursor', config.setCursor);

  builder.setCursor(config.cursor);
  builder.setSelect(config.select);

  builder.addScale({
    scaleKey: 'x',
    isTime: false,
    distribution: ScaleDistribution.Ordinal,
    orientation: xOri,
    direction: xDir,
  });

  builder.addAxis({
    scaleKey: 'x',
    isTime: false,
    placement: xOri === 0 ? AxisPlacement.Bottom : AxisPlacement.Left,
    splits: config.xSplits,
    values: config.xValues,
    grid: false,
    ticks: false,
    gap: 15,
    theme,
  });

  let seriesIndex = 0;

  // iterate the y values
  for (let i = 1; i < data.fields.length; i++) {
    const field = data.fields[i];

    field.state!.seriesIndex = seriesIndex++;

    const customConfig: BarChartFieldConfig = { ...defaultBarChartFieldConfig, ...field.config.custom };

    const scaleKey = field.config.unit || FIXED_UNIT;
    const colorMode = getFieldColorModeForField(field);
    const scaleColor = getFieldSeriesColor(field, theme);
    const seriesColor = scaleColor.color;

    builder.addSeries({
      scaleKey,
      pxAlign: false,
      lineWidth: customConfig.lineWidth,
      lineColor: seriesColor,
      //lineStyle: customConfig.lineStyle,
      fillOpacity: customConfig.fillOpacity,
      theme,
      colorMode,
      pathBuilder: config.drawBars,
      pointsBuilder: config.drawPoints,
      show: !customConfig.hideFrom?.graph,
      gradientMode: customConfig.gradientMode,
      thresholds: field.config.thresholds,

      // The following properties are not used in the uPlot config, but are utilized as transport for legend config
      dataFrameFieldIndex: {
        fieldIndex: i,
        frameIndex: 0,
      },
      fieldName: getFieldDisplayName(field, data),
      hideInLegend: customConfig.hideFrom?.legend,
    });

    // The builder will manage unique scaleKeys and combine where appropriate
    builder.addScale({
      scaleKey,
      min: field.config.min,
      max: field.config.max,
      softMin: customConfig.axisSoftMin,
      softMax: customConfig.axisSoftMax,
      orientation: yOri,
      direction: yDir,
    });

    if (customConfig.axisPlacement !== AxisPlacement.Hidden) {
      let placement = customConfig.axisPlacement;
      if (!placement || placement === AxisPlacement.Auto) {
        placement = AxisPlacement.Left;
      }
      if (xOri === 1) {
        if (placement === AxisPlacement.Left) {
          placement = AxisPlacement.Bottom;
        }
        if (placement === AxisPlacement.Right) {
          placement = AxisPlacement.Top;
        }
      }

      builder.addAxis({
        scaleKey,
        label: customConfig.axisLabel,
        size: customConfig.axisWidth,
        placement,
        formatValue: (v) => formattedValueToString(field.display!(v)),
        theme,
      });
    }
  }

  return builder;
}

/** @internal */
export function preparePlotFrame(data: DataFrame[]) {
  const firstFrame = data[0];
  const firstString = firstFrame.fields.find((f) => f.type === FieldType.string);

  if (!firstString) {
    throw new Error('No string field in DF');
  }

  const resultFrame = new MutableDataFrame();
  resultFrame.addField(firstString);

  for (const f of firstFrame.fields) {
    if (f.type === FieldType.number) {
      resultFrame.addField(f);
    }
  }

  return resultFrame;
}
