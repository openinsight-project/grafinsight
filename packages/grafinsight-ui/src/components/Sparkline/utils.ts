import { DataFrame, FieldConfig, FieldSparkline, IndexVector } from '@grafinsight/data';
import { GraphFieldConfig } from '../uPlot/config';

/** @internal
 * Given a sparkline config returns a DataFrame ready to be turned into Plot data set
 **/
export function preparePlotFrame(sparkline: FieldSparkline, config?: FieldConfig<GraphFieldConfig>): DataFrame {
  const length = sparkline.y.values.length;
  const yFieldConfig = {
    ...sparkline.y.config,
    ...config,
  };

  return {
    refId: 'sparkline',
    fields: [
      sparkline.x ?? IndexVector.newField(length),
      {
        ...sparkline.y,
        config: yFieldConfig,
      },
    ],
    length,
  };
}
