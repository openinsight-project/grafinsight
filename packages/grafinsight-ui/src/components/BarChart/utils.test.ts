import { preparePlotConfigBuilder, preparePlotFrame } from './utils';
import { FieldConfig, FieldType, GrafInsightTheme, MutableDataFrame, VizOrientation } from '@grafinsight/data';
import { BarChartFieldConfig, BarChartOptions, BarStackingMode, BarValueVisibility } from './types';
import { GraphGradientMode } from '../uPlot/config';
import { LegendDisplayMode } from '../VizLegend/types';

function mockDataFrame() {
  const df1 = new MutableDataFrame({
    refId: 'A',
    fields: [{ name: 'ts', type: FieldType.string, values: ['a', 'b', 'c'] }],
  });

  const df2 = new MutableDataFrame({
    refId: 'B',
    fields: [{ name: 'ts', type: FieldType.time, values: [1, 2, 4] }],
  });

  const f1Config: FieldConfig<BarChartFieldConfig> = {
    displayName: 'Metric 1',
    decimals: 2,
    unit: 'm/s',
    custom: {
      gradientMode: GraphGradientMode.Opacity,
      lineWidth: 2,
      fillOpacity: 0.1,
    },
  };

  const f2Config: FieldConfig<BarChartFieldConfig> = {
    displayName: 'Metric 2',
    decimals: 2,
    unit: 'kWh',
    custom: {
      gradientMode: GraphGradientMode.Hue,
      lineWidth: 2,
      fillOpacity: 0.1,
    },
  };

  df1.addField({
    name: 'metric1',
    type: FieldType.number,
    config: f1Config,
    state: {},
  });

  df2.addField({
    name: 'metric2',
    type: FieldType.number,
    config: f2Config,
    state: {},
  });

  return preparePlotFrame([df1, df2]);
}

describe('GraphNG utils', () => {
  describe('preparePlotConfigBuilder', () => {
    const frame = mockDataFrame();

    const config: BarChartOptions = {
      orientation: VizOrientation.Auto,
      groupWidth: 20,
      barWidth: 2,
      showValue: BarValueVisibility.Always,
      legend: {
        displayMode: LegendDisplayMode.List,
        placement: 'bottom',
        calcs: [],
      },
      stacking: BarStackingMode.None,
    };

    it.each([VizOrientation.Auto, VizOrientation.Horizontal, VizOrientation.Vertical])('orientation', (v) => {
      expect(
        preparePlotConfigBuilder(frame!, { colors: { panelBg: '#000000' } } as GrafInsightTheme, {
          ...config,
          orientation: v,
        })
      ).toMatchSnapshot();
    });

    it.each([BarValueVisibility.Always, BarValueVisibility.Auto])('value visibility', (v) => {
      expect(
        preparePlotConfigBuilder(frame!, { colors: { panelBg: '#000000' } } as GrafInsightTheme, {
          ...config,
          showValue: v,
        })
      ).toMatchSnapshot();
    });

    it.each([BarStackingMode.None, BarStackingMode.Percent, BarStackingMode.Standard])('stacking', (v) => {
      expect(
        preparePlotConfigBuilder(frame!, { colors: { panelBg: '#000000' } } as GrafInsightTheme, {
          ...config,
          stacking: v,
        })
      ).toMatchSnapshot();
    });
  });
});
