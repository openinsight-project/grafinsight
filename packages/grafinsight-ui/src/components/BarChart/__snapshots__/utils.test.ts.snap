// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphNG utils preparePlotConfigBuilder orientation 1`] = `
UPlotConfigBuilder {
  "axes": Object {
    "m/s": UPlotAxisBuilder {
      "props": Object {
        "formatValue": [Function],
        "label": undefined,
        "placement": "bottom",
        "scaleKey": "m/s",
        "size": undefined,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
      },
    },
    "x": UPlotAxisBuilder {
      "props": Object {
        "gap": 15,
        "grid": false,
        "isTime": false,
        "placement": "left",
        "scaleKey": "x",
        "splits": [Function],
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "ticks": false,
        "values": [Function],
      },
    },
  },
  "bands": Array [],
  "cursor": Object {
    "points": Object {
      "show": false,
    },
    "x": false,
    "y": false,
  },
  "getTimeZone": [Function],
  "hasBottomAxis": true,
  "hasLeftAxis": true,
  "hooks": Object {
    "drawClear": Array [
      [Function],
    ],
    "init": Array [
      [Function],
    ],
    "setCursor": Array [
      [Function],
    ],
  },
  "scales": Array [
    UPlotScaleBuilder {
      "props": Object {
        "direction": -1,
        "distribution": "ordinal",
        "isTime": false,
        "orientation": 1,
        "scaleKey": "x",
      },
    },
    UPlotScaleBuilder {
      "props": Object {
        "direction": 1,
        "max": undefined,
        "min": undefined,
        "orientation": 0,
        "scaleKey": "m/s",
        "softMax": undefined,
        "softMin": 0,
      },
    },
  ],
  "select": Object {
    "show": false,
  },
  "series": Array [
    UPlotSeriesBuilder {
      "props": Object {
        "colorMode": Object {
          "description": "Derive colors from thresholds",
          "getCalculator": [Function],
          "id": "thresholds",
          "isByValue": true,
          "name": "From thresholds",
        },
        "dataFrameFieldIndex": Object {
          "fieldIndex": 1,
          "frameIndex": 0,
        },
        "fieldName": "Metric 1",
        "fillOpacity": 0.1,
        "gradientMode": "opacity",
        "hideInLegend": undefined,
        "lineColor": "#808080",
        "lineWidth": 2,
        "pathBuilder": [Function],
        "pointsBuilder": [Function],
        "pxAlign": false,
        "scaleKey": "m/s",
        "show": true,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "thresholds": undefined,
      },
    },
  ],
  "tzDate": [Function],
}
`;

exports[`GraphNG utils preparePlotConfigBuilder orientation 2`] = `
UPlotConfigBuilder {
  "axes": Object {
    "m/s": UPlotAxisBuilder {
      "props": Object {
        "formatValue": [Function],
        "label": undefined,
        "placement": "bottom",
        "scaleKey": "m/s",
        "size": undefined,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
      },
    },
    "x": UPlotAxisBuilder {
      "props": Object {
        "gap": 15,
        "grid": false,
        "isTime": false,
        "placement": "left",
        "scaleKey": "x",
        "splits": [Function],
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "ticks": false,
        "values": [Function],
      },
    },
  },
  "bands": Array [],
  "cursor": Object {
    "points": Object {
      "show": false,
    },
    "x": false,
    "y": false,
  },
  "getTimeZone": [Function],
  "hasBottomAxis": true,
  "hasLeftAxis": true,
  "hooks": Object {
    "drawClear": Array [
      [Function],
    ],
    "init": Array [
      [Function],
    ],
    "setCursor": Array [
      [Function],
    ],
  },
  "scales": Array [
    UPlotScaleBuilder {
      "props": Object {
        "direction": -1,
        "distribution": "ordinal",
        "isTime": false,
        "orientation": 1,
        "scaleKey": "x",
      },
    },
    UPlotScaleBuilder {
      "props": Object {
        "direction": 1,
        "max": undefined,
        "min": undefined,
        "orientation": 0,
        "scaleKey": "m/s",
        "softMax": undefined,
        "softMin": 0,
      },
    },
  ],
  "select": Object {
    "show": false,
  },
  "series": Array [
    UPlotSeriesBuilder {
      "props": Object {
        "colorMode": Object {
          "description": "Derive colors from thresholds",
          "getCalculator": [Function],
          "id": "thresholds",
          "isByValue": true,
          "name": "From thresholds",
        },
        "dataFrameFieldIndex": Object {
          "fieldIndex": 1,
          "frameIndex": 0,
        },
        "fieldName": "Metric 1",
        "fillOpacity": 0.1,
        "gradientMode": "opacity",
        "hideInLegend": undefined,
        "lineColor": "#808080",
        "lineWidth": 2,
        "pathBuilder": [Function],
        "pointsBuilder": [Function],
        "pxAlign": false,
        "scaleKey": "m/s",
        "show": true,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "thresholds": undefined,
      },
    },
  ],
  "tzDate": [Function],
}
`;

exports[`GraphNG utils preparePlotConfigBuilder orientation 3`] = `
UPlotConfigBuilder {
  "axes": Object {
    "m/s": UPlotAxisBuilder {
      "props": Object {
        "formatValue": [Function],
        "label": undefined,
        "placement": "left",
        "scaleKey": "m/s",
        "size": undefined,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
      },
    },
    "x": UPlotAxisBuilder {
      "props": Object {
        "gap": 15,
        "grid": false,
        "isTime": false,
        "placement": "bottom",
        "scaleKey": "x",
        "splits": [Function],
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "ticks": false,
        "values": [Function],
      },
    },
  },
  "bands": Array [],
  "cursor": Object {
    "points": Object {
      "show": false,
    },
    "x": false,
    "y": false,
  },
  "getTimeZone": [Function],
  "hasBottomAxis": true,
  "hasLeftAxis": true,
  "hooks": Object {
    "drawClear": Array [
      [Function],
    ],
    "init": Array [
      [Function],
    ],
    "setCursor": Array [
      [Function],
    ],
  },
  "scales": Array [
    UPlotScaleBuilder {
      "props": Object {
        "direction": 1,
        "distribution": "ordinal",
        "isTime": false,
        "orientation": 0,
        "scaleKey": "x",
      },
    },
    UPlotScaleBuilder {
      "props": Object {
        "direction": 1,
        "max": undefined,
        "min": undefined,
        "orientation": 1,
        "scaleKey": "m/s",
        "softMax": undefined,
        "softMin": 0,
      },
    },
  ],
  "select": Object {
    "show": false,
  },
  "series": Array [
    UPlotSeriesBuilder {
      "props": Object {
        "colorMode": Object {
          "description": "Derive colors from thresholds",
          "getCalculator": [Function],
          "id": "thresholds",
          "isByValue": true,
          "name": "From thresholds",
        },
        "dataFrameFieldIndex": Object {
          "fieldIndex": 1,
          "frameIndex": 0,
        },
        "fieldName": "Metric 1",
        "fillOpacity": 0.1,
        "gradientMode": "opacity",
        "hideInLegend": undefined,
        "lineColor": "#808080",
        "lineWidth": 2,
        "pathBuilder": [Function],
        "pointsBuilder": [Function],
        "pxAlign": false,
        "scaleKey": "m/s",
        "show": true,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "thresholds": undefined,
      },
    },
  ],
  "tzDate": [Function],
}
`;

exports[`GraphNG utils preparePlotConfigBuilder stacking 1`] = `
UPlotConfigBuilder {
  "axes": Object {
    "m/s": UPlotAxisBuilder {
      "props": Object {
        "formatValue": [Function],
        "label": undefined,
        "placement": "bottom",
        "scaleKey": "m/s",
        "size": undefined,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
      },
    },
    "x": UPlotAxisBuilder {
      "props": Object {
        "gap": 15,
        "grid": false,
        "isTime": false,
        "placement": "left",
        "scaleKey": "x",
        "splits": [Function],
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "ticks": false,
        "values": [Function],
      },
    },
  },
  "bands": Array [],
  "cursor": Object {
    "points": Object {
      "show": false,
    },
    "x": false,
    "y": false,
  },
  "getTimeZone": [Function],
  "hasBottomAxis": true,
  "hasLeftAxis": true,
  "hooks": Object {
    "drawClear": Array [
      [Function],
    ],
    "init": Array [
      [Function],
    ],
    "setCursor": Array [
      [Function],
    ],
  },
  "scales": Array [
    UPlotScaleBuilder {
      "props": Object {
        "direction": -1,
        "distribution": "ordinal",
        "isTime": false,
        "orientation": 1,
        "scaleKey": "x",
      },
    },
    UPlotScaleBuilder {
      "props": Object {
        "direction": 1,
        "max": undefined,
        "min": undefined,
        "orientation": 0,
        "scaleKey": "m/s",
        "softMax": undefined,
        "softMin": 0,
      },
    },
  ],
  "select": Object {
    "show": false,
  },
  "series": Array [
    UPlotSeriesBuilder {
      "props": Object {
        "colorMode": Object {
          "description": "Derive colors from thresholds",
          "getCalculator": [Function],
          "id": "thresholds",
          "isByValue": true,
          "name": "From thresholds",
        },
        "dataFrameFieldIndex": Object {
          "fieldIndex": 1,
          "frameIndex": 0,
        },
        "fieldName": "Metric 1",
        "fillOpacity": 0.1,
        "gradientMode": "opacity",
        "hideInLegend": undefined,
        "lineColor": "#808080",
        "lineWidth": 2,
        "pathBuilder": [Function],
        "pointsBuilder": [Function],
        "pxAlign": false,
        "scaleKey": "m/s",
        "show": true,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "thresholds": undefined,
      },
    },
  ],
  "tzDate": [Function],
}
`;

exports[`GraphNG utils preparePlotConfigBuilder stacking 2`] = `
UPlotConfigBuilder {
  "axes": Object {
    "m/s": UPlotAxisBuilder {
      "props": Object {
        "formatValue": [Function],
        "label": undefined,
        "placement": "bottom",
        "scaleKey": "m/s",
        "size": undefined,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
      },
    },
    "x": UPlotAxisBuilder {
      "props": Object {
        "gap": 15,
        "grid": false,
        "isTime": false,
        "placement": "left",
        "scaleKey": "x",
        "splits": [Function],
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "ticks": false,
        "values": [Function],
      },
    },
  },
  "bands": Array [],
  "cursor": Object {
    "points": Object {
      "show": false,
    },
    "x": false,
    "y": false,
  },
  "getTimeZone": [Function],
  "hasBottomAxis": true,
  "hasLeftAxis": true,
  "hooks": Object {
    "drawClear": Array [
      [Function],
    ],
    "init": Array [
      [Function],
    ],
    "setCursor": Array [
      [Function],
    ],
  },
  "scales": Array [
    UPlotScaleBuilder {
      "props": Object {
        "direction": -1,
        "distribution": "ordinal",
        "isTime": false,
        "orientation": 1,
        "scaleKey": "x",
      },
    },
    UPlotScaleBuilder {
      "props": Object {
        "direction": 1,
        "max": undefined,
        "min": undefined,
        "orientation": 0,
        "scaleKey": "m/s",
        "softMax": undefined,
        "softMin": 0,
      },
    },
  ],
  "select": Object {
    "show": false,
  },
  "series": Array [
    UPlotSeriesBuilder {
      "props": Object {
        "colorMode": Object {
          "description": "Derive colors from thresholds",
          "getCalculator": [Function],
          "id": "thresholds",
          "isByValue": true,
          "name": "From thresholds",
        },
        "dataFrameFieldIndex": Object {
          "fieldIndex": 1,
          "frameIndex": 0,
        },
        "fieldName": "Metric 1",
        "fillOpacity": 0.1,
        "gradientMode": "opacity",
        "hideInLegend": undefined,
        "lineColor": "#808080",
        "lineWidth": 2,
        "pathBuilder": [Function],
        "pointsBuilder": [Function],
        "pxAlign": false,
        "scaleKey": "m/s",
        "show": true,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "thresholds": undefined,
      },
    },
  ],
  "tzDate": [Function],
}
`;

exports[`GraphNG utils preparePlotConfigBuilder stacking 3`] = `
UPlotConfigBuilder {
  "axes": Object {
    "m/s": UPlotAxisBuilder {
      "props": Object {
        "formatValue": [Function],
        "label": undefined,
        "placement": "bottom",
        "scaleKey": "m/s",
        "size": undefined,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
      },
    },
    "x": UPlotAxisBuilder {
      "props": Object {
        "gap": 15,
        "grid": false,
        "isTime": false,
        "placement": "left",
        "scaleKey": "x",
        "splits": [Function],
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "ticks": false,
        "values": [Function],
      },
    },
  },
  "bands": Array [],
  "cursor": Object {
    "points": Object {
      "show": false,
    },
    "x": false,
    "y": false,
  },
  "getTimeZone": [Function],
  "hasBottomAxis": true,
  "hasLeftAxis": true,
  "hooks": Object {
    "drawClear": Array [
      [Function],
    ],
    "init": Array [
      [Function],
    ],
    "setCursor": Array [
      [Function],
    ],
  },
  "scales": Array [
    UPlotScaleBuilder {
      "props": Object {
        "direction": -1,
        "distribution": "ordinal",
        "isTime": false,
        "orientation": 1,
        "scaleKey": "x",
      },
    },
    UPlotScaleBuilder {
      "props": Object {
        "direction": 1,
        "max": undefined,
        "min": undefined,
        "orientation": 0,
        "scaleKey": "m/s",
        "softMax": undefined,
        "softMin": 0,
      },
    },
  ],
  "select": Object {
    "show": false,
  },
  "series": Array [
    UPlotSeriesBuilder {
      "props": Object {
        "colorMode": Object {
          "description": "Derive colors from thresholds",
          "getCalculator": [Function],
          "id": "thresholds",
          "isByValue": true,
          "name": "From thresholds",
        },
        "dataFrameFieldIndex": Object {
          "fieldIndex": 1,
          "frameIndex": 0,
        },
        "fieldName": "Metric 1",
        "fillOpacity": 0.1,
        "gradientMode": "opacity",
        "hideInLegend": undefined,
        "lineColor": "#808080",
        "lineWidth": 2,
        "pathBuilder": [Function],
        "pointsBuilder": [Function],
        "pxAlign": false,
        "scaleKey": "m/s",
        "show": true,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "thresholds": undefined,
      },
    },
  ],
  "tzDate": [Function],
}
`;

exports[`GraphNG utils preparePlotConfigBuilder value visibility 1`] = `
UPlotConfigBuilder {
  "axes": Object {
    "m/s": UPlotAxisBuilder {
      "props": Object {
        "formatValue": [Function],
        "label": undefined,
        "placement": "bottom",
        "scaleKey": "m/s",
        "size": undefined,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
      },
    },
    "x": UPlotAxisBuilder {
      "props": Object {
        "gap": 15,
        "grid": false,
        "isTime": false,
        "placement": "left",
        "scaleKey": "x",
        "splits": [Function],
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "ticks": false,
        "values": [Function],
      },
    },
  },
  "bands": Array [],
  "cursor": Object {
    "points": Object {
      "show": false,
    },
    "x": false,
    "y": false,
  },
  "getTimeZone": [Function],
  "hasBottomAxis": true,
  "hasLeftAxis": true,
  "hooks": Object {
    "drawClear": Array [
      [Function],
    ],
    "init": Array [
      [Function],
    ],
    "setCursor": Array [
      [Function],
    ],
  },
  "scales": Array [
    UPlotScaleBuilder {
      "props": Object {
        "direction": -1,
        "distribution": "ordinal",
        "isTime": false,
        "orientation": 1,
        "scaleKey": "x",
      },
    },
    UPlotScaleBuilder {
      "props": Object {
        "direction": 1,
        "max": undefined,
        "min": undefined,
        "orientation": 0,
        "scaleKey": "m/s",
        "softMax": undefined,
        "softMin": 0,
      },
    },
  ],
  "select": Object {
    "show": false,
  },
  "series": Array [
    UPlotSeriesBuilder {
      "props": Object {
        "colorMode": Object {
          "description": "Derive colors from thresholds",
          "getCalculator": [Function],
          "id": "thresholds",
          "isByValue": true,
          "name": "From thresholds",
        },
        "dataFrameFieldIndex": Object {
          "fieldIndex": 1,
          "frameIndex": 0,
        },
        "fieldName": "Metric 1",
        "fillOpacity": 0.1,
        "gradientMode": "opacity",
        "hideInLegend": undefined,
        "lineColor": "#808080",
        "lineWidth": 2,
        "pathBuilder": [Function],
        "pointsBuilder": [Function],
        "pxAlign": false,
        "scaleKey": "m/s",
        "show": true,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "thresholds": undefined,
      },
    },
  ],
  "tzDate": [Function],
}
`;

exports[`GraphNG utils preparePlotConfigBuilder value visibility 2`] = `
UPlotConfigBuilder {
  "axes": Object {
    "m/s": UPlotAxisBuilder {
      "props": Object {
        "formatValue": [Function],
        "label": undefined,
        "placement": "bottom",
        "scaleKey": "m/s",
        "size": undefined,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
      },
    },
    "x": UPlotAxisBuilder {
      "props": Object {
        "gap": 15,
        "grid": false,
        "isTime": false,
        "placement": "left",
        "scaleKey": "x",
        "splits": [Function],
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "ticks": false,
        "values": [Function],
      },
    },
  },
  "bands": Array [],
  "cursor": Object {
    "points": Object {
      "show": false,
    },
    "x": false,
    "y": false,
  },
  "getTimeZone": [Function],
  "hasBottomAxis": true,
  "hasLeftAxis": true,
  "hooks": Object {
    "drawClear": Array [
      [Function],
    ],
    "init": Array [
      [Function],
    ],
    "setCursor": Array [
      [Function],
    ],
  },
  "scales": Array [
    UPlotScaleBuilder {
      "props": Object {
        "direction": -1,
        "distribution": "ordinal",
        "isTime": false,
        "orientation": 1,
        "scaleKey": "x",
      },
    },
    UPlotScaleBuilder {
      "props": Object {
        "direction": 1,
        "max": undefined,
        "min": undefined,
        "orientation": 0,
        "scaleKey": "m/s",
        "softMax": undefined,
        "softMin": 0,
      },
    },
  ],
  "select": Object {
    "show": false,
  },
  "series": Array [
    UPlotSeriesBuilder {
      "props": Object {
        "colorMode": Object {
          "description": "Derive colors from thresholds",
          "getCalculator": [Function],
          "id": "thresholds",
          "isByValue": true,
          "name": "From thresholds",
        },
        "dataFrameFieldIndex": Object {
          "fieldIndex": 1,
          "frameIndex": 0,
        },
        "fieldName": "Metric 1",
        "fillOpacity": 0.1,
        "gradientMode": "opacity",
        "hideInLegend": undefined,
        "lineColor": "#808080",
        "lineWidth": 2,
        "pathBuilder": [Function],
        "pointsBuilder": [Function],
        "pxAlign": false,
        "scaleKey": "m/s",
        "show": true,
        "theme": Object {
          "colors": Object {
            "panelBg": "#000000",
          },
        },
        "thresholds": undefined,
      },
    },
  ],
  "tzDate": [Function],
}
`;
